### PR TITLE
update get_asset_url function

### PR DESCRIPTION
**Pull request by karosc**
This pull request addresses #1527 by adding and if statement to the `get_asset_url` function in dash.py instructing the app to use the external asset path if it is defined. 

I have tested this with an short I wrote and [published](https://github.com/karosc/Dash_Tester). If you run the app and examine the src attribute of the image, it references and image on my own site karosc.github.io.

### CODE REVIEW

**Comment by alexcjohnson** [comment]
Very nice! Could be a little DRYer, as it's only the first arg to `get_asset_path` that needs to change, but that's not a big deal.

Can you add a couple of unit tests of `app.get_asset_url`? Should cover both the local and external cases, make an app and test a couple of asset urls for that app, but we don't need to run the app (so the external url need not be valid). `test_configs.py` looks like a good place to put it, and it would look something like these two:

https://github.com/plotly/dash/blob/3402cb8f67466bcee8776b1d07dd3134d08bda76/tests/unit/test_configs.py#L147-L179

After that, we'll just need a changelog entry and it'll be ready to go!

**Comment by karosc** [comment]
does this seem more DRY? Not sure exactly what you were referring to.

```python
def get_asset_url(self, path):
        if self.config.assets_external_path:
            prefix = self.config.assets_external_path
        else:
	    prefix = self.config.requests_pathname_prefix

        asset = get_asset_path(
                     prefix,
                     path,
                     self.config.assets_url_path.lstrip("/"),
               )

        return asset
```

**Comment by alexcjohnson** [comment]
Yep, that's about what I had in mind. But again, not a big deal.

**Comment by karosc** [comment]
After building some tests, I am realizing that the third argument also needs to change. Conceptually `assets_external_path` is the full external path to the assets folder ([assuming this is the case based on the code here](https://github.com/plotly/dash/blob/3402cb8f67466bcee8776b1d07dd3134d08bda76/dash/dash.py#L1101), where the resource url is created by combining assets_external_path only with the asset path and not also with the asset folder path). 

Based on this, the function should look more like:

```python

    def get_asset_url(self, path):
        if self.config.assets_external_path:
            asset = get_asset_path(
                self.config.assets_external_path,
                path,
                '',
            )
        else:
            asset = get_asset_path(
                self.config.requests_pathname_prefix,
                path,
                self.config.assets_url_path.lstrip("/"),
            )

        return asset

```

File: tests/unit/test_configs.py [hunk 1]
```diff
@@ -105,6 +105,28 @@ def test_pathname_prefix_assets(empty_environ, req, expected):
...
 
+@pytest.mark.parametrize(
+    "requests_pathname_prefix, assets_external_path, assets_url_path, expected",
+    [
+        (None    , None                        , "assets" , "/assets/reset.css"),
+        ("/app/" , None                        , "assets" , '/app/assets/reset.css'),
+        (None    , None                        , "css"    , "/css/reset.css"),
+        ("/app/" , None                        , "css"    , "/app/css/reset.css"),
+        (None    , "http:\\external.com/", "assets" , "http:\\external.com/assets/reset.css"),
+        ("/app/" , "http:\\external.com/", "css"    , "http:\\external.com/css/reset.css"),
```
Comment (alexcjohnson) [inline tests/unit/test_configs.py:116]:
```suggestion
        (None    , "http://external.com/", "assets" , "http://external.com/assets/reset.css"),
        ("/app/" , "http://external.com/", "css"    , "http://external.com/css/reset.css"),
```

File: dash/dash.py [hunk 2]
```diff
@@ -1098,8 +1100,9 @@ def _setup_server(self):
     def _add_assets_resource(self, url_path, file_path):
         res = {"asset_path": url_path, "filepath": file_path}
         if self.config.assets_external_path:
-            res["external_url"] = "{}{}".format(
-                self.config.assets_external_path, url_path
+            res["external_url"] = "/".join([
+                self.config.assets_external_path.rstrip('/'), self.config.assets_url_path.strip('/'), url_path.lstrip('/')
+            ]
```
Comment (alexcjohnson) [inline dash/dash.py:1105]: Hmm, maybe we could just call `get_asset_url` here?

**Comment by alexcjohnson** [comment]
Looking good, just a couple of minor comments above, but also I think it needs linting (I sympathize with wanting aligned values in the test matrix but I don't think `black` will go for it 😅 ) - Probably just needs `npm run format`.

File: dash/dash.py (hunk 2 above)
Comment (karosc) [reply dash/dash.py]: That works! Passing all tests.

**Comment by karosc** [comment]
running npm format changed a bunch of other files, sorry for creating more review work...🤷‍♂️

**Comment by alexcjohnson** [comment]
Ah, I'm guessing you're on a newer version of `black`. What version do you have installed? We've talked about the need to update to 20.8b1 (and maybe now is the right time to do that across Dash, cc @Marc-Andre-Rivet?) but currently we have 19.10b0

https://github.com/plotly/dash/blob/c5ba38f0ae7b7f8c173bda10b4a8ddd035f1d867/requires-dev.txt#L11

**Comment by karosc** [comment]
oh shoot, you're right I am running 20.8b1. I can downgrade and reformat then if that's when you want to do.

**Comment by alexcjohnson** [comment]
Downgrading and reformatting would be the easiest way to get this merged quickly 😄 We may upgrade but that's going to take a little work because we'll want to do it evenly across all the dash repos.

**Comment by karosc** [comment]
Alright, I reset to before I ran the format routine, downgraded black to 19.10b0 and then redid the format. Things came out to be more expected since only my changes were reformatted by black.

**Review by alexcjohnson — APPROVED** [review]
Very nice - thank you! 💃

**Pull request merged**